ultimarc/devices/mini_pac.py

```diff
@@ -62,14 +64,80 @@ class MiniPacDevice(USBDeviceHandle):
     class_descr = _('Mini-PAC')
     interface = 2
 
+    def to_json_str(self, pac_struct):
+        """ Converts a PacStruct to a json object """
+        json_obj = {'schemaVersion': 2.0, 'resourceType': 'mini-pac-pins', 'deviceClass': self.class_id}
+
+        # debounce
+        # TODO - fully implement debounce behavior
+        json_obj['debounce'] = 'standard'
+
+        # macros
+        macros = []
+        macro_start = 0xe0
+        macro_index = 1
+
+        y = 0
+        for x in range(MACRO_START_INDEX, len(pac_struct.bytes)):
+            if x >= y:
+                macro = {}
+                if pac_struct.bytes[x]:
+                    if pac_struct.bytes[x] == macro_start:
+                        macro['name'] = f'macro #{macro_index}'
+                        macro_start += 1
+                        macro_index += 1
+
+                        action = []
+                        for y in range (x+1, len(pac_struct.bytes)):
+                            # check that the value isn't zero and not the start of the next macro
+                            if pac_struct.bytes[y] and pac_struct.bytes[y] != macro_start:
+                                action.append(get_ipac_series_mapping_key(pac_struct.bytes[y]))
+                            else:
+                                macro['action'] = action
+                                macros.append(macro)
+                                break
+                else:
+                    # No more macros defined, end the loop
+                    break
+        if len(macros):
+            json_obj['macros'] = macros
+
+        # pins
+        pins = []
+        for key in PinMapping:
+            action_index, alternate_action_index, shift_index = PinMapping[key]
+            pin = {}
+            if pac_struct.bytes[action_index]:
+                pin['name'] = key
+                pin['action'] = get_ipac_series_mapping_key(pac_struct.bytes[action_index])
+                if pac_struct.bytes[alternate_action_index]:
+                    pin['alternate_action'] = get_ipac_series_mapping_key(pac_struct.bytes[alternate_action_index])
+                if pac_struct.bytes[shift_index]:
+                    pin['shift'] = True
+                pins.append(pin)
+        json_obj['pins'] = pins
+        return json_obj
```

Comment:
Seems like this would be a good spot to validate the JSON dict against the schema json to validate that the JSON created here is valid.

Reply:
Added validation